File: src/main/java/no/nav/personopplysninger/features/personalia/PersonaliaRestConfiguration.java
package no.nav.personopplysninger.features.personalia;

import com.fasterxml.jackson.databind.ObjectMapper;
import no.nav.personopplysninger.features.ConsumerFactory;
import no.nav.security.oidc.jaxrs.OidcClientRequestFilter;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import javax.inject.Named;
import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.ClientRequestFilter;
import javax.ws.rs.ext.ContextResolver;
import java.net.URI;
import java.net.URISyntaxException;

@Configuration
public class PersonaliaRestConfiguration {

    @Value("${PERSONOPPLYSNINGER_API_TPS_PROXY_API_V1_INNSYN_APIKEY_PASSWORD}")
    private String tpsProxyApiKeyPassword;

    @Value("${PERSONOPPLYSNINGER_API_DKIF_API_APIKEY_PASSWORD}")
    private String dkifApiKeyPassword;

    @Bean
    public PersonConsumer personConsumer(
            @Named("tpsProxyClient") Client client,
            @Value("${TPS_PROXY_API_V1_INNSYN_URL}") String personServiceUri) throws URISyntaxException {
        return new PersonConsumer(client, new URI(personServiceUri));
    }

    @Bean
    public KontaktinfoConsumer kontaktinformasjonConsumer(
            @Named("dkifClient") Client client,
            @Value("${DKIF_API_URL}") String kontaktinfoServiceUri) throws URISyntaxException {
        return new KontaktinfoConsumer(client, new URI(kontaktinfoServiceUri));
    }

    @Bean
    public Client tpsProxyClient(ContextResolver<ObjectMapper> clientObjectMapperResolver) {
        return clientBuilder(clientObjectMapperResolver)
                .register((ClientRequestFilter) requestContext -> requestContext.getHeaders()
                        .putSingle(ConsumerFactory.DEFAULT_APIKEY_USERNAME, tpsProxyApiKeyPassword))
                .build();
    }

    @Bean
    public Client dkifClient(ContextResolver<ObjectMapper> clientObjectMapperResolver) {
        return clientBuilder(clientObjectMapperResolver)
                .register((ClientRequestFilter) requestContext -> requestContext.getHeaders()
                        .putSingle(ConsumerFactory.DEFAULT_APIKEY_USERNAME, dkifApiKeyPassword))
                .build();
    }

    private ClientBuilder clientBuilder(ContextResolver<ObjectMapper> clientObjectMapperResolver) {
        return ClientBuilder.newBuilder()
                .register(OidcClientRequestFilter.class)
                .register(clientObjectMapperResolver);
    }
}

File: src/main/kotlin/no/nav/personopplysninger/features/personalia/PersonaliaRestConfiguration.kt
package no.nav.personopplysninger.features.personalia

import com.fasterxml.jackson.databind.ObjectMapper
import no.nav.personopplysninger.features.personalia.pdl.PdlConsumer
import no.nav.personopplysninger.features.tokendings.TokenDingsService
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.net.URI
import java.net.URISyntaxException
import javax.inject.Named
import javax.ws.rs.client.Client
import javax.ws.rs.client.ClientBuilder
import javax.ws.rs.ext.ContextResolver

@Configuration
open class PersonaliaRestConfiguration {

    @Value("\${DIGDIR_KRR_PROXY_CONSUMER_TARGET_APP}")
    private val digdirTargetApp: String? = null

    @Value("\${PDL_CONSUMER_TARGET_APP}")
    private val pdlTargetApp: String? = null

    @Bean
    @Throws(URISyntaxException::class)
    open fun personConsumer(
        @Named("tpsProxyClient") client: Client,
        @Value("\${TPS_PROXY_API_V1_INNSYN_URL}") personServiceUri: String
    ): PersonConsumer {
        return PersonConsumer(client, URI(personServiceUri))
    }

    @Bean
    @Throws(URISyntaxException::class)
    open fun kontaktinformasjonConsumer(
        @Named("digdirKrrClient") client: Client,
        @Value("\${DIGDIR_KRR_PROXY_URL}") kontaktinfoServiceUri: String,
        tokenDingsService: TokenDingsService
    ): KontaktinfoConsumer {
        return KontaktinfoConsumer(client, URI(kontaktinfoServiceUri), tokenDingsService, digdirTargetApp)
    }

    @Bean
    open fun tpsProxyClient(clientObjectMapperResolver: ContextResolver<ObjectMapper>): Client {
        return clientBuilder(clientObjectMapperResolver).build()
    }

    @Bean
    @Throws(URISyntaxException::class)
    open fun pdlConsumer(
        @Named("pdlCLient") client: Client,
        @Value("\${PDL_API_URL}") pdlUri: String,
        tokenDingsService: TokenDingsService
    ): PdlConsumer {
        return PdlConsumer(client, URI(pdlUri), tokenDingsService, pdlTargetApp)
    }

    @Bean
    open fun pdlCLient(clientObjectMapperResolver: ContextResolver<ObjectMapper>): Client {
        return clientBuilder(clientObjectMapperResolver).build()
    }

    @Bean
    open fun digdirKrrClient(clientObjectMapperResolver: ContextResolver<ObjectMapper>): Client {
        return clientBuilder(clientObjectMapperResolver).build()
    }

    private fun clientBuilder(clientObjectMapperResolver: ContextResolver<ObjectMapper>): ClientBuilder {
        return ClientBuilder.newBuilder().register(clientObjectMapperResolver)
    }
}
